Octree/include/CGAL/Octree/Node.h

```diff
@@ -0,0 +1,346 @@
+
+    return (*m_children)[index];
+  }
+
+  /*!
+   * \brief Read-only access the child nodes of this node by their indices
+   *
+   * \param index The index of the child node, as an int
+   * \return A const reference to the node
+   */
+  const Node<Point_index> &operator[](int index) const {
+
+    assert(!is_leaf());
+    assert(0 <= index && index < 8);
+
+    return (*m_children)[index];
+  }
+
+  /// @}
+
+  /// \name Property Accessors
+  /// @{
+
+  /*!
+   * \brief Read-only access to this node's parent
+   *
+   * Ownership of a node is not equivalent to ownership of the entire tree,
+   * so it's not possible to obtain write access to a node's parent,
+   * only its children.
+   * Note that the return type is nullable,
+   * attempting to find the parent of a root node will return null.
+   * \todo Should I instead assert the node isn't root? (that would make this undefined behavior)
+   *
+   * \return A const pointer to the parent of this node (possibly nullptr)
+   */
+  const Node<Point_index> *parent() const { return m_parent; }
+
+  /*!
+   * \brief Retrieve this node's depth in the tree
+   * \return the depth of this node, where root has a depth of 0
+   */
+  const uint8_t &depth() const { return m_depth; }
+
+  /*!
+   * \brief Retrieve this node's location in the tree
+   *
+   * \todo Should I link to an explanation of the location type?
+   *
+   * \return this node's location
+   */
+  Int_location location() const {
+
+    return m_location;
+  }
+
+  /*!
+   * \brief Retrieve this node's index in relation to its parent
+   * \return the index of this nod3
+   */
+  Index index() const {
+
+    // TODO: There must be a better way of doing this!
+
+    Index result;
+
+    result[0] = location()[0] & 1;
+    result[1] = location()[1] & 1;
+    result[2] = location()[2] & 1;
+
+    return result;
+  }
+
+  /*!
+   * \brief Determine whether this node is a leaf node
+   * \return whether this node has no children
+   */
+  bool is_leaf() const { return (!m_children); }
+
+  /*!
+   * \brief Determine whether this node is the root node
+   * \return whether this node has no parent
+   */
+  bool is_root() const { return (!m_parent); }
+
+  /// @}
+
+  /// \name Value Accessors
+  /// @{
+
+  /*!
+   * \brief Access to the content held by this node
+   * \return a reference to the collection of point indices
+   */
+  Point_range &points() { return m_points; }
+
+  /*!
+   * \brief Read-only access to the content held by this node
+   * \return a read-only reference to the collection of point indices
+   */
+  const Point_range &points() const { return m_points; }
+
+  /*!
+   * \brief Check whether this node contains any points
+   * \return if this node contains no points
+   */
+  bool is_empty() const {
+    return m_points.empty();
+  }
+
+  /*!
+   * \brief Count the points contained by this node
+   * \return the number of points this node owns
+   */
+  std::size_t number_of_points() const {
```

Comment:
Maybe just `size()`?

Reply:
I used `number_of_points` to match [Point_set_3](https://doc.cgal.org/latest/Point_set_3/classCGAL_1_1Point__set__3.html)'s equivalent function (the same reason I used `is_empty`), but it occurs to me that I probably don't need to be so explicit since the octree doesn't have "removed" points.

Reply:
A small subtlety here is that `Point_set_3` models the range concept (which means [all the functions of the range are available while not directly documented](https://doc.cgal.org/latest/Circulator/classRange.html)) and was also written based on `Surface_mesh` which explains why:

- `empty()` and `size()` are available (model of Range)
- `is_empty()` and `number_of_points()` are available (consistency with Surface Mesh) and basically do the same thing

In your case, I think it would actually be a good idea to make Node a model of Range too. You probably already have all the necessary functions (again, see the link above): you have `begin() / end()` for the points, you have the size, the empty test, etc.

And in that case, I would remove `number_of_points()`and just use `size()` (same for `empty()` and not `is_empty()`), as you don't really need consistency with `Point_set_3` or `Surface_mesh`.

Reply:
Good idea! If I make the Node class satisfy the range concept, I should probably store a `m_begin` and `m_end` iterators rather than storing a `PointRange` like I'm doing now.